Inline images

xqwzts opened the pull request:
Closes #13 

I've changed the way `img` elements get parsed by the builder, removing them from the list of block elements.
Previously the code assumed that we would only have `TextSpan`s as valid inline elements, so I changed the inline children list type to `List<Object>` to accept both `TextSpan` and `Image` widgets.
Finally I wrapped all the inline objects in a `Wrap` widget before placing it in the tree, to control the layout of text and images.

This still has some annoying text layout issues, but feels like a better alternative to the current situation.

DaveShuckerow commented:
This is a non-trivial change.  I'll take a detailed look at this tomorrow.

DaveShuckerow commented:
I'm reading the code and I'm curious: the inline merging used to create a single TextSpan, which AFAIK propagates style changes like **bold** or *italic* down to the span's children.  Now that you're breaking this up when you find an image element inline, 

Does *this still ![image](https://user-images.githubusercontent.com/14947366/32524330-1089b05c-c3d4-11e7-96c8-f914c43b8c4e.png) work*?

If not, let's talk about the design to structure the node visitor so that this still does work.

xqwzts commented:
Ah good catch, that actually throws [here](https://github.com/flutter/flutter_markdown/pull/14/files#diff-ce56fad35df83a521bce930d5a008630R198) while trying to add the Image as a child of the `em` TextSpan.

So in this case we have `em -> [TextSpan, Image, TextSpan]`, but I hadn't taken into account the case where an `Image` is a child of another text element.

I think this can be resolved with the same logic that was added in `_addAnonymousBlockIfNeeded`, I'll take a quick stab at it. Would changing the visitor be a cleaner way of doing things?

xqwzts commented:
Updated with the fix.

DaveShuckerow commented on lib/src/builder.dart:
```diff
@@ -138,68 +137,96 @@ class MarkdownBuilder implements md.NodeVisitor {
+        children.add(_buildImage(element.attributes['src']));
+      } else {
+        if (current.children.isNotEmpty) {
+          List<TextSpan> spans = <TextSpan>[];
```
Really, the underlying problem is that this code was all designed around the idea that TextSpan and InlineElement were roughly equivalent.  Now, multiple types of object can be inline, so we need to rethink the way this works.

The purpose of using TextSpans isn't to represent all of the text at once, but rather to style sections of the text correctly.

However, the parse tree of Markdown should already provide us with this information: it should know if we're in a **bolded section** 

## or a header section
or any other type of section.

So, I think the best approach is to skip the TextSpan stage.  Inline sections already need to be widgets that we stitch together next to each other, so it should be just as OK to put Texts next to each other as it is to put Texts next to Images next to other Texts.

This will save you the trouble of translating an element into a TextSpan or an Image, and then having to translate again from TextSpan to Text or passing the Image through

xqwzts replied:
@DaveShuckerow Thanks for taking the time, I haven't had much time to really dig into it until now. The TextSpans give us two advantages:

1- Word wrapping instead of widget wrapping. We lose this with any implementation that doesn't just merge TextSpans.
2- Inherited styles: so "something like _this **would** work_". Is there a way to merge TextStyles to achieve this without having a TextSpan for each style?


> However, the parse tree of Markdown should already provide us with this information:

Good point, while we don't have access to the whole tree in the visitor, we can keep track of the tag of each InlineElement:

```dart
class _InlineElement {
  _InlineElement(this.tag);

  final String tag;
  final List<Widget> children = <Widget>[];
}
```


And in `visitText` style the current text based on its tag, and the tags of any other inlineElements it may be nested within [and directly create a `RichText` widget to contain it]

```dart
void visitText(md.Text text) {
  
   ...

    TextSpan styledSpan = span;
    if (_inlines.length > 1) {
      for (int i = _inlines.length-1 ; i >= 0; i--) {
        TextSpan styleWrapper = new TextSpan(style: styleSheet.styles[_inlines[i].tag], children: [styledSpan]);
        styledSpan = styleWrapper;
      }
    }

    _inlines.last.children.add(new RichText(text: styledSpan,));
}
```

This is definitely cleaner than a list of Objects, and looping to determine if each one is a text element or an image. I have a couple of todos* to get through before pushing, will hopefully get that done in a day.

*- changing InlineElement to have a single child widget instead of a list of children [we aren't merging children anymore using TextSpans so each inlineElement is its own widget]
*- chasing an edge case bug with sentences beginning with an inline element not being nested in a block.

While I feel like the change makes the code cleaner and better, it still relies on multiple nested `TextSpan`s just to inherit styling, and fails point 1 above related to word wrapping.

DaveShuckerow replied:
1: If Wrap() doesn't properly wrap texts over each other, then you can merge texts together like so:
```dart
var mergedTexts = <Widget>[];
for (var i=0; i<current.children.lengh; i++) {
  if (mergedTexts.last is RichText && current.children[i] is RichText) {
    // Replace mergedTexts.last with a new RichText containing the TextSpans inside of both of them
  } else {
    mergedTexts.add(current.children[i]);
  }
}
current.children.clear();
current.children.addAll(mergedTexts);
```

This way you guarantee that inline elements will only have widget children.

2: [TextStyle.copyWith](https://docs.flutter.io/flutter/painting/TextStyle/copyWith.html) should be enough for copying a text's style without nesting TextSpans.

Each InlineElement would be created with its style copied from its parent:

```dart
var parentStyle = _inlines.last?.style ?? new TextStyle();
_inlines.add(new _InlineElement(tag, style: parentStyle.merge(styleSheet.styles[tag]));
```

This way you can persist textstyles across inline elements, regardless of whether or not they're a textspan.

xqwzts commented:
Updated and rebased.

@DaveShuckerow `TextStyle.merge` was key, thanks for pointing it out.

InlineElements now get styled according to their parents [inlines and blocks].
RichText widgets are created in `visitText`.
Adjacent RichText widgets are merged in `_addAnonymousBlockIfNeeded`

The output is flatter too without all the nested TextSpans, which had the added benefit of making the tests easier to read

DaveShuckerow commented on lib/src/builder.dart:
```diff
@@ -39,7 +38,11 @@ class _BlockElement {
 
 class _InlineElement {
-  final List<TextSpan> children = <TextSpan>[];
+  _InlineElement(this.tag, {this.style});
```
Dartdoc isn't usually necessary on private classes, but in this case, the behavior of this class is not obvious.

Summarize the discussion we had in our review to explain what this does.

xqwzts replied:
Done.

DaveShuckerow commented on lib/src/builder.dart:
```diff
@@ -119,7 +131,17 @@ class MarkdownBuilder implements md.NodeVisitor {
       _blocks.add(new _BlockElement(tag));
     } else {
-      _inlines.add(new _InlineElement());
+       if (_inlines.isEmpty) {
```
This is the same thing as line 107.  Can you refactor it out into a helper method `_addParentInlineIfNeeded`?

xqwzts replied:
Done.

xqwzts commented:
Travis is failing on the image tests which use a dummy 'img' value instead of a real uri, this commit https://github.com/flutter/flutter/commit/5fc8eb8263801c17c360eaf189e0f6b94bfdc686 added a throw when the response isn't valid, which is what we're hitting.

I've tried replacing 'img' with the flutter logo url 'https://flutter.io/images/flutter-mark-square-100.png', but that also throws a 404, is some extra configuration needed to grant the test runner network access?

DaveShuckerow commented:
>  is some extra configuration needed to grant the test runner network access?

I'll look into this and make sure that we get the tests passing again.  Thanks for the heads-up!

DaveShuckerow commented:
NetworkImage should be able to run in Travis tests.  Travis doesn't restrict network access in any way.  I was able to reproduce the failure locally on my development machine.

I filed https://github.com/flutter/flutter/issues/13433 tracking this issue.

xqwzts commented:
I think we might need to use a MockHttpClient the way the net tests do: https://github.com/flutter/flutter/blob/28366002d9a86ec1403764453df31580362c9097/packages/flutter_tools/test/base/net_test.dart will take a look at that in a bit.

xqwzts commented:
> I appreciate your patience on this change. It's a very non-trivial refactoring, and I want to make sure that we do it right so that we don't have to do it a second time :).

Thanks for the guidance, this is much better than what I started out with :)

xqwzts commented:
I've opened a separate PR #17 for the broken tests.

xqwzts commented:
Rebased.

xqwzts commented:
Thanks for being patient and guiding me through it!
